### PR TITLE
Improve logging on election timeout

### PR DESCRIPTION
**Pull request by npepinpe**
## Description

This PR improves logging when an election occurs due to missing heartbeats. This fixes the issue of having very high "time since last heartbeat" due when there are no heartbeats, and it also shows that there is no known leader instead of just showing null.

This also exposes the election timeout to display it, making a difference between the time out and when the last heartbeat was received.

Previously, the log message would look like this when there was no leader and no heartbeat had ever been sent:

```
2021-06-20 14:53:50.739 [] [raft-server-2-raft-partition-partition-1] INFO 
      io.atomix.raft.roles.FollowerRole - RaftServer{raft-partition-partition-1}{role=FOLLOWER} - No heartbeat from null in the last 1624200830739ms, sending poll requests
```

Now it will look like this:

```
2021-06-20 14:53:50.739 [] [raft-server-2-raft-partition-partition-1] INFO 
      io.atomix.raft.roles.FollowerRole - RaftServer{raft-partition-partition-1}{role=FOLLOWER} - No heartbeat from a known leader since 2603ms
```

Additionally, the sending poll requests part was split off, and when we send poll requests we will now log the following. If there are other known active members to poll:

```
10:59:21.175 [] [raft-server-1-raft-partition-partition-1] INFO
      io.atomix.raft.roles.FollowerRole - RaftServer{raft-partition-partition-1}{role=FOLLOWER} - Sending poll requests to all active members: [DefaultRaftMember{id=0, type=ACTIVE, updated=2021-06-23T08:59:18.475Z}, DefaultRaftMember{id=2, type=ACTIVE, updated=2021-06-23T08:59:18.475Z}]
```

If there are no other known active members to poll:

```
2021-06-20 14:53:50.739 [] [raft-server-2-raft-partition-partition-1] INFO 
      io.atomix.raft.roles.FollowerRole - RaftServer{raft-partition-partition-1}{role=FOLLOWER} - Transitioning to candidate as there are no known other active members
```

Finally, in Raft, we log every single request/response sent as debug. This creates tons of noise as we are sending lots and lots of requests. This PR lowers the log level of this to trace. We already log relevant debug information when handling polls/appends/etc., so it's pretty redundant.

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

File: atomix/cluster/src/main/java/io/atomix/raft/roles/FollowerRole.java [hunk 1]
```diff
@@ -243,11 +245,18 @@ private void schedulePollRequests() {
...
+              .map(MemberId::id)
+              .orElse("a known leader");
+
       log.info(
-          "No heartbeat from {} in the last {}ms, sending poll requests",
-          raft.getLeader(),
-          missTime);
+          "No heartbeat from {} after {} (time since last heartbeat in millis: {})",
+          leader,
+          electionTimer.getElectionTimeout(),
```
[the remark this answers is not in the document]
Comment (npepinpe) [reply atomix/cluster/src/main/java/io/atomix/raft/roles/FollowerRole.java:257]: My reasoning was they may not always be the same, and both are interesting information. When diagnosing issues for a cluster we don't operate, it can be useful to see the timeout being used (could be useful to spot bugs, for example), as well as the last time the heartbeat was received (which can be older than the last election time out, but if newer could indicate some bug in our scheduling). It can also inform us of issues, e.g. overloaded thread context queue, if the last heartbeat was a while ago but it took us quite some time to actually do the election. Since the timeout is typically randomized and we don't always know, it's hard to tell if there are major discrepancies.

That said, in the common case, they're probably the same, so I wouldn't insist here to keep it. If you feel it's just noise, we can remove it.

[the remark this answers is not in the document]
Comment (deepthidevaki) [reply atomix/cluster/src/main/java/io/atomix/raft/roles/FollowerRole.java:257]: In that case, I would rephrase it. 
` "No heartbeat from {} for that last {currentTime - lastHeartbeatTime} ms, election timeout is {}ms)"`

We can also initialize `lastHearbeatTime` to` System.currentTimeMillis()` in the constructor, so we don't need the check.

Also this only logs the configured election timeout, which is not really useful. We can see it from the startup logs. If you want to see the actual randomized delay, then electionTimer.getElectionTimeout() should return the current delay instead of the configured `electionTimeout`.

[the remark this answers is not in the document]
Comment (npepinpe) [reply atomix/cluster/src/main/java/io/atomix/raft/roles/FollowerRole.java:257]: Initializing it is definitely better, can't believe I didn't think of that :sweat_smile: I don't think it's particularly useful to distinguish if we have ever received one or not, so let's do that. And as for the election timeout, then my mistake, I thought the reset actually changed the election time out and that's what was used for the delay, but I admit I mostly just glanced at it and assumed it worked like this :sweat_smile:

**Comment by npepinpe** [comment]
I updated it and decided to move the election timeout log as a trace level log in the RandomizedElectionTimer, mostly because the fact it changes is implementation detail, and as a timeout could potentially be small, it would be called very often and create lots of noise.

**Review by deepthidevaki — APPROVED** [review]
:rocket:

**Comment by npepinpe** [comment]
bors merge
